File: LeetCode/19. Remove Nth Node from End of List/Solution.java
/**
 * Given a linked list, remove the n-th node from the end of list and return its head.
 * 
 * Example:
 * Given linked list: 1->2->3->4->5, and n = 2.
 * 
 * After removing the second node from the end, the linked list becomes 1->2->3->5.
 * Note:
 * Given n will always be valid.
 *
 * Follow up:
 *
 * Could you do this in one pass?
 */

 /**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null)
            return head;
        
        ListNode slow = head;
        ListNode fast = head;
        
        int count = 0;
        while(count < n) {
            fast = fast.next;
            count++;
        }
        
        if (fast == null)
            return head.next;
        
        while(fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }
        
        slow.next = slow.next.next;
        return head;
    }
}

File: LeetCode/19. Remove Nth Node from End of List/Solution.cpp
/**
 * Given a linked list, remove the n-th node from the end of list and return its head.
 * 
 * Example:
 * Given linked list: 1->2->3->4->5, and n = 2.
 * 
 * After removing the second node from the end, the linked list becomes 1->2->3->5.
 * Note:
 * Given n will always be valid.
 *
 * Follow up:
 *
 * Could you do this in one pass?
 */

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        if (head == NULL || n < 1)
            return NULL;
        
        ListNode dummy(-1);
        dummy.next = head;
        
        ListNode* lead = &dummy;
        ListNode* trail = &dummy;
        
        int count = 0;
        while (count < n)
        {
            lead = lead->next;
            count++;
        }
        
        while(lead->next != NULL)
        {
            trail = trail->next;
            lead = lead->next;
        }
        
        trail->next = trail->next->next;
        return dummy.next;

    }
};